shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/routeDetailsPage/RouteBranching.kt

```diff
@@ -0,0 +1,691 @@
+                    stopGraph.putEdge(d to e)
+                    stopGraph.removeEdge(d to b)
+                    stopGraph.putEdge(e to f)
+                    stopGraph.removeEdge(b to f)
+                    stopGraph.removeEdge(b to a)
+                }
+            }
+            Pair("Boat-F7", 1) -> {
+                // as of 2025-07-14 morning routes visit Logan only after Central Wharf and
+                // afternoon routes visit Logan only before Seaport/Fan so Logan never gets
+                // disambiguated
+                val a = DisambiguatedStopId("Boat-Quincy", 0)
+                val b = DisambiguatedStopId("Boat-Logan", 0)
+                val c = DisambiguatedStopId("Boat-Fan", 0)
+                val d = DisambiguatedStopId("Boat-Aquarium", 0)
+                val e = DisambiguatedStopId("Boat-Logan", 1)
+                val f = DisambiguatedStopId("Boat-Quincy", 1)
+                if (stopGraph.vertices.keys == setOf(a, b, c, d, f)) {
+                    stopGraph.putVertex(e, checkNotNull(stopGraph.vertices[b]))
+                    stopGraph.putEdge(d to e)
+                    stopGraph.removeEdge(d to b)
+                    stopGraph.putEdge(e to f)
+                    stopGraph.removeEdge(b to f)
+                    stopGraph.removeEdge(b to a)
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Encapsulates the logic that turns A->B->D->E->F typical B->C->E atypical into
+ *
+ * ```text
+ *    ┍ A
+ *   ┌┾ B
+ * ® │┝ C
+ *   ┥│ D
+ *   └┾ E
+ *    ┕ F
+ * ```
+ *
+ * (although the actual Unicode graph rendering is in ProjectUtils).
+ *
+ * A _segment_ in this context is a list of stops visited consecutively that are either all typical
+ * or all non-typical. A segment boundary is drawn where
+ * - stops change from typical to non-typical or non-typical to typical
+ * - there are multiple next stops
+ * - the next stop has multiple previous stops
+ *
+ * It is worth noting that a segment boundary is specifically not drawn if
+ * - stops change from one non-typical typicality to another
+ * - some trips are truncated versions of other similarly typical trips
+ */
+data class RouteBranching(
+    /**
+     * Records the connections between individual stops, preserving how typical stops and
+     * connections are.
+     */
+    val stopGraph: Graph<DisambiguatedStopId, Pair<Stop, Set<Typicality>>>?,
+    /** Records the connections between segments. */
+    val segmentGraph: Graph<DisambiguatedStopId, BranchSegmentInGraph>?,
+    /** The final list of segments to draw in order. */
+    val segments: List<BranchSegment>?,
+) {
+    enum class StickSide {
+        Left,
+        Right;
+
+        operator fun unaryMinus() =
+            when (this) {
+                Left -> Right
+                Right -> Left
+            }
+    }
+
+    data class StickSideState(
+        val before: Boolean,
+        val currentStop: Boolean,
```

Comment:
suggestion(non-blocking): It took me until I read where this is set in code to understand `currentStop` - some docs here might help.